src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -151,16 +164,46 @@ case class WorkflowActor(workflow: WorkflowDescriptor, backend: Backend)
     case (k: CallKey, v) if k.scope == key.scope && isShard(k) => (k, v)
   }
 
+  /** Always returns stay(), but messages self to perform the requested transition once all persistence is complete. */
+  private def transition(toState: WorkflowState): State = {
+    def handleTerminalWorkflow: Future[Unit] = {
+      for {
+        _ <- backend.cleanUpForWorkflow(workflow)
+        _ <- globalDataAccess.updateWorkflowOptions(workflow.id, workflow.workflowOptions.clearEncryptedValues)
+        _ = self ! Terminate
+      } yield ()
+    }
+
+    if (stateName != toState) {
+      val transitionFuture = for {
+      // Write the new workflow state before logging the change, tests assume the change is in effect when
+      // the message is logged.
+        _ <- globalDataAccess.updateWorkflowState(workflow.id, toState)
+        _ = logger.info(s"Beginning transition from $stateName to $toState.")
+        _ <- if (toState.isTerminal) handleTerminalWorkflow else Future.successful(())
+      } yield ()
+
+      transitionFuture recover {
+        case e: Exception => logger.error(s"Failed to transition workflow status from $stateName to $toState", e)
+      }
+
+      // Only message self to perform the transition after the persistence of the new state is complete.
+      transitionFuture map { _ => PerformTransition(toState) } pipeTo self
+    }
+
+    stay()
```

Comment:
Yeah it kinda confuses me when I see it too.  :smile:  How about a different name like `scheduleTransition` to clarify that this will not be a synchronous `goto()`?